### PR TITLE
[staging] linux.stdenv: gcc_10 → gcc_11

### PR DESCRIPTION
**Pull request by fabianhjr**
###### Description of changes

Was able to build and boot into gnome 42 under kernel 5.17, seems like most packages already support gcc11 without changes; only 1 package gave issues and pinning it to a previous version was enough.

```
cat /proc/version                                                                                                                                                                                                                                                       Thu 24 Mar 2022 09:42:13 PM CST
Linux version 5.17.0 (nixbld@localhost) (gcc (GCC) 11.2.0, GNU ld (GNU Binutils) 2.35.2) #1-NixOS SMP PREEMPT Sun Mar 20 20:14:17 UTC 2022
```

Currently running a `home-manager switch` but those are a ton more (~4500) over the system required packages (~2600)

EDIT:

Additionally built `linux_4_9` (oldest supported LTS)

EDIT2:

Home environment also built, there are some packages that need updating but this works great so far! :3

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

### CODE REVIEW

**Comment by fabianhjr** [comment]
@ofborg eval

**Review by kevincox — APPROVED** [review]

**Comment by vcunat** [comment]
Oh, this was quite fast.  I'd expect that most of porting work [would be C++ stuff](https://gcc.gnu.org/gcc-11/porting_to.html).  I'll hope that it will go fine enough to go through the usual staging route, without a dedicated jobset.  It's not so new, so patches for anything important should be easy to find.

**Comment by fabianhjr** [comment]
Yeah, I started my local build with GCC11 thinking there would be a ton of blockers and next day with only 1 patch I was already booted. (The home-manager switch was almost successful too, had to add 3 more patches)

**Comment by trofi** [comment]
Worth moving `gfortran` to 11 version as well?

**Comment by fabianhjr** [comment]
> Worth moving `gfortran` to 11 version as well?

Probably, but didn't test; it is not necessary afaik.

**Comment by fabianhjr** [comment]
Taking a look, building lastest staging since binutils-2.38 got merged and seems like a linking/ld issue.

**Comment by fabianhjr** [comment]
Was able to reproduce locally on current staging, investigating the cause and possible fixes.

**Comment by fabianhjr** [comment]
@mweinelt, seems like the change of c++14 to c++17 by default changed abseil but not grpc so there was a link time issue. The following patch fixes the build on nixpkgs but will need to take a look at packages that depend on abseil.

```
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 16704f0ec64..f5fb9db5bae 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -17225,7 +17225,12 @@ with pkgs;
 
   grilo-plugins = callPackage ../development/libraries/grilo-plugins { };
 
-  grpc = callPackage ../development/libraries/grpc { };
+  grpc = callPackage ../development/libraries/grpc {
+    # grpc builds with c++14 so abseil must also be built that way
+    abseil-cpp = abseil-cpp.override {
+      cxxStandard = "14";
+    };
+  };
 
   gsettings-qt = libsForQt5.callPackage ../development/libraries/gsettings-qt { };
```

File: pkgs/top-level/all-packages.nix
Comment (vcunat) [inline pkgs/top-level/all-packages.nix:12141]: Why adding yet another exception instead of changing the default 10 below?

Comment (fabianhjr) [reply pkgs/top-level/all-packages.nix:12141]: I have no access to x86_64-darwin nor aarch64-linux platforms.

PR was built and tested locally on x86_64-linux before proposing it alongsides binutils 2.38 bump that also got into staging. (Which had issues on x86_64-darwin while x86_64-linux was working fine at the time; I though that binutils bump was also only going to happen for x86_64-linux)

Comment (vcunat) [reply pkgs/top-level/all-packages.nix:12141]: But the default does not apply to those platforms.

Comment (vcunat) [reply pkgs/top-level/all-packages.nix:12141]: Well, darwin doesn't use gcc as the default compiler, but maybe for `x86_64-darwin` there is some difference here.

Comment (fabianhjr) [reply pkgs/top-level/all-packages.nix:12141]: The only GCC version not pinned would be x86_64-darwin. If a @NixOS/darwin-maintainers has bandwidth to bump GCC on that platform then that route seems better but don't want to push on a platform I neither use nor can help with.

Comment (vcunat) [reply pkgs/top-level/all-packages.nix:12141]: I had i686-linux in mind mainly, and the overall complexity of the switch.  Anyway, thanks for explanation.  I'm looking into this.

**Comment by samuela** [comment]
After this change we should also update https://github.com/NixOS/nixpkgs/blob/1d63f89caaf140ac33f1796dc0aaeb20f4ac4e55/pkgs/development/compilers/cudatoolkit/versions.toml#L43-L59. cc @NixOS/cuda-maintainers

File: pkgs/top-level/all-packages.nix
Comment (vcunat) [reply pkgs/top-level/all-packages.nix:12141]: Hmm, I'll probably leave it this way now and maybe come back once it reaches master.

Comment (uri-canva) [reply pkgs/top-level/all-packages.nix:12141]: I've tested `x86_64-darwin` and it seems ok: https://github.com/NixOS/nixpkgs/pull/168093. Unfortunately it times out on ofborg.

**Comment by samuela** [comment]
This change broke dm-tree, jax, optax, https://github.com/NixOS/nixpkgs/issues/169134, https://github.com/NixOS/nixpkgs/issues/169148, flax, and certainly many many others. What is the plan for doing this migration? How should packages respond when this broke their build?

I suspect the total number of breakages to be in the hundreds.

**Comment by fabianhjr** [comment]
There are several paths; currently doing as many fixes as possible but at the rate of 10 or so per day; if a specific gcc version is needed then an override of the stdenv would do. (`stdenv = if stdenv.cc.isGNU then gcc10Stdenv else stdenv`)

**Comment by fabianhjr** [comment]
google-cloud-cpp seems to be related to a mismatch between abseil `-std` and google-cloud-cpp `-std` a similar override to abseil-cpp as that done on grpc should fix it. Currently working on a merge request.

**Comment by fabianhjr** [comment]
google-cloud-cpp fix: https://github.com/NixOS/nixpkgs/pull/169220

Can't reproduce python3Packages.jaxlib build issue.

**Comment by fabianhjr** [comment]
I would recommend either attempting a fix or if not possible opening an issue and notifying the maintainers.

**Comment by samuela** [comment]
> I would recommend either attempting a fix or if not possible opening an issue and notifying the maintainers.

I am the maintainer :P

**Comment by fabianhjr** [comment]
Probably better to open a separate issue to discuss possible solutions since what you point out (testing in staging-next before merging and block merging in case of breakage) might be lost / not reach other stakeholders on this merge request.

**Comment by samuela** [comment]
Yes, we're seeing a significant number of breakages every time staging/staging-next get merged. It consumes an enormous amount of maintainer time and energy for us to play "catch up" with these massive changes that break wide swaths of our package landscape.

**Comment by vcunat** [comment]
That's "the same" everywhere.  Only for packages built on Hydra usually someone else is fixing the regressions.

**Comment by samuela** [comment]
This just doesn't seem like a sustainable long-term development model to me. As a maintainer it seems like I'm forced to either:
* Constantly track all of the changes happening on master and recognize breakages before they make it into master. Maintainers have no infrastructure or support in doing this. My packages aren't in any hydra jobsets. How am I supposed to know when people break things before those changes end up in master?
* Try to unbreak massive breakages after staging gets merged in and everything is messed up for reasons that maintainers don't understand or have context on. To make matters worse, there's 900 commits to sift through. So good luck sorting through the mess... This is a huge burden and leads to burnout. 

Neither of these options are sustainable IMHO. We need a better system. Personally, I don't think people should be able to push PRs breaking other people's packages without a fallback option and a stated plan for completing a migration to the new version. Otherwise it's just chaos.

**Comment by vcunat** [comment]
Quite many breakages do reach master all the time, even for packages built on Hydra, if we look at absolute numbers.  It's one of the reasons why stable releases exist, also in many other distros.  If you care about packages that don't get build on Hydra.nixos.org and/or OfBorg CI, I can't see how to significantly improve this for them without attacking the fact that they don't get build.  Without that noone will even know that there is a breakage.  (we're getting very off-topic on this thread)

In general, this gcc bump in particular had larger negative impact than I hoped, but that doesn't seem very related to the "system" questions.

**Comment by fabianhjr** [comment]
In retrospective I was expecting a larger staging-next cycle of 2 weeks but it ended up being close to 3 days. (1 day when everything was building and only 2 days of fixes before it hit the main branch)

Even then I would have probably missed fixing breakages of packages / variants not built by hydra.

**Comment by samuela** [comment]
> In particular, this gcc bump had larger negative impact than I hoped, but that doesn't seem very related to the "system" questions.

Agreed, we should take that discussion to a different thread somewhere. maybe discourse?

> In retrospective I was expecting a larger staging-next cycle of 2 weeks but it ended up being close to 3 days. (1 day when everything was building and only 2 days of fixes before it hit the main branch)

Indeed and thank you for taking this on and addressing the failures so far, @fabianhjr! GCC bumps are not for the faint of heart, but they are important work. I don't fault you personally in this quagmire; I think the current development model, which I can only summarize as "branch aggressively and cross your fingers" is lacking for everyone involved. We need to think about designing a better system to give everyone more confidence in being able to make large changes more safely.

**Comment by vcunat** [comment]
We do stabilization in separate branches and jobsets for some risky changes.

**Comment by samuela** [comment]
I started a discussion on Discourse for the broader conversation: https://discourse.nixos.org/t/nixpkgss-current-development-workflow-is-not-sustainable/18741.

**Comment by trofi** [comment]
rili broke, proposed fix: https://github.com/NixOS/nixpkgs/pull/169540

**Comment by trofi** [comment]
`zkfuse` broke, proposed fix: https://github.com/NixOS/nixpkgs/pull/169543

File: pkgs/top-level/all-packages.nix
Comment (vcunat) [reply pkgs/top-level/all-packages.nix:12141]: OK, `wine` wouldn't build if I switched i686 to gcc11 as well, with errors during `mingw-w64-i686-w64-mingw32-9.0.0.drv`, and aarch64-linux WIP is separate in #167726.  So I guess the rest of the mess will remain for now.

**Pull request merged**
